akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -651,6 +670,25 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    */
   final protected def fail[T](out: Outlet[T], ex: Throwable): Unit = interpreter.fail(conn(out), ex)
 
+  /**
+   * Automatically invokes [[cancel]] or [[complete]] on all the input or output ports that have been called,
+   * then marks the stage as stopped.
+   */
+  final def cancelStage(): Unit =
```

Comment:
is this adding API that is then different in master? mima problem?

Reply:
That's right, this doesn't work as a new public method. I guess the easiest way would be to make this method work exactly like in 2.6 and add a `cause` parameter which is ignored in 2.5.

Reply:
The other option would be to make it `private[stream]` so that it is only ever called from the default `OutHandler.onDownstreamFinish` implementation.

However, that would prevent custom stage implementations to use `cancelStage`, so they couldn't use the new propagation modes.

Reply:
Implemented the first option.